tests/test_elements/test_mol2mol/test_mol2mol_filter.py

```diff
@@ -14,68 +22,172 @@
+            "c1ccccc1": (1, None),
+            "Cl": 1,
+        }
+        smiles_filter = SmilesFilter(smiles_pats)
+
+        for filter_ in [smarts_filter, smiles_filter]:
+            new_input_as_list = list(filter_.patterns.keys())
+            pipeline = Pipeline(
+                [
+                    ("Smiles2Mol", SmilesToMol()),
+                    ("SmartsFilter", filter_),
+                    ("Mol2Smiles", MolToSmiles()),
+                    ("ErrorFilter", ErrorFilter()),
+                ],
+            )
+            filtered_smiles = pipeline.fit_transform(SMILES_LIST)
+            self.assertEqual(
+                filtered_smiles, [SMILES_BENZENE, SMILES_CHLOROBENZENE, SMILES_CL_BR]
+            )
+
+            pipeline.set_params(SmartsFilter__keep_matches=False)
+            filtered_smiles_2 = pipeline.fit_transform(SMILES_LIST)
+            self.assertEqual(filtered_smiles_2, [SMILES_ANTIMONY, SMILES_METAL_AU])
+
+            pipeline.set_params(
+                SmartsFilter__mode="all", SmartsFilter__keep_matches=True
+            )
+            filtered_smiles_3 = pipeline.fit_transform(SMILES_LIST)
+            self.assertEqual(filtered_smiles_3, [SMILES_CHLOROBENZENE])
+
+            pipeline.set_params(
+                SmartsFilter__keep_matches=True, SmartsFilter__patterns=["I"]
+            )
+            filtered_smiles_4 = pipeline.fit_transform(SMILES_LIST)
+            self.assertEqual(filtered_smiles_4, [])
+
+            pipeline.set_params(
+                SmartsFilter__keep_matches=False,
+                SmartsFilter__mode="any",
+                SmartsFilter__patterns=new_input_as_list,
+            )
+            filtered_smiles_5 = pipeline.fit_transform(SMILES_LIST)
+            self.assertEqual(filtered_smiles_5, [SMILES_ANTIMONY, SMILES_METAL_AU])
+
+    def test_smarts_filter_parallel(self) -> None:
+        """Test if molecules are filtered correctly by allowed SMARTS patterns in parallel."""
+        smarts_pats: dict[str, Union[int, tuple[Optional[int], Optional[int]]]] = {
+            "c": (4, None),
+            "Cl": 1,
+            "cc": (1, None),
+            "ccc": (1, None),
+            "cccc": (1, None),
+            "ccccc": (1, None),
+            "cccccc": (1, None),
+            "c1ccccc1": (1, None),
+            "cCl": 1,
+        }
+        smarts_filter = SmartsFilter(smarts_pats, mode="all", n_jobs=-1)
+        pipeline = Pipeline(
+            [
+                ("Smiles2Mol", SmilesToMol()),
+                ("SmartsFilter", smarts_filter),
+                ("Mol2Smiles", MolToSmiles()),
+                ("ErrorFilter", ErrorFilter()),
             ],
         )
-        filtered_smiles = pipeline.fit_transform(
+        filtered_smiles = pipeline.fit_transform(SMILES_LIST)
+        self.assertEqual(filtered_smiles, [SMILES_CHLOROBENZENE])
+
+    def test_descriptor_filter(self) -> None:
```

Comment:
Since the DescriptorFilter filters numeric values, including floats, a test that checks the bounds would be essential. So, you should test that for some descriptor value $x$ you get the expected results when you set the filter value $f$ to $f=x$, $f=x - \epsilon$, $f=x + \epsilon$. A typical bug in such code is that the equals case is handled differently than expected.

Reply:
I see the point. However, I am not quite sure for which property this makes sense to test. I can take something like MolWt, HOWEVER, this is always a crooked value, so I am not sure how to generate an "equal" cas. 

But if I take something like NumHAcceptors, this is always an int. So there I can test the boundaries for the equal case (also use floats as boundaries). I will implement this, but please check, whether this is actually what you meant...

Reply:
Already looks good. But you should also add a case with the exact value (the equals case). In your example you should also test 2.00 in addition to 1.99 and 2.01. For example:

```python
pipeline.set_params(
       DescriptorsFilter__descriptors={
            "NumHAcceptors": (2.00, 4),
       }
 )
```

The "equals case" is a classical edge case because you have to decide in your implementation whether you want to make it an including or excluding interval.

Reply:
also included this